include/ghex/unstructured/pattern.hpp

```diff
@@ -0,0 +1,340 @@
+/* 
+ * GridTools
+ * 
+ * Copyright (c) 2014-2020, ETH Zurich
+ * All rights reserved.
+ * 
+ * Please, refer to the LICENSE file in the root directory.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * 
+ */
+#ifndef INCLUDED_GHEX_UNSTRUCTURED_PATTERN_HPP
+#define INCLUDED_GHEX_UNSTRUCTURED_PATTERN_HPP
+
+#include <vector>
+#include <map>
+#include <numeric>
+#include <cstring>
+#include <iosfwd>
+
+#include "../transport_layer/context.hpp"
+#include "../allocator/unified_memory_allocator.hpp"
+#include "../pattern.hpp"
+#include "../buffer_info.hpp"
+#include "./grid.hpp"
+
+
+namespace gridtools {
+    namespace ghex {
+
+        /** @brief unstructured pattern specialization
+         *
+         * This class provides access to the receive and send iteration spaces, determined by the halos,
+         * and holds all connections to the neighbors.
+         *
+         * @tparam Communicator communicator type
+         * @tparam Index index type for domain and iteration space
+         * @tparam DomainId domain id type*/
+        template<typename Communicator, typename Index, typename DomainId>
+        class pattern<Communicator, unstructured::detail::grid<Index>, DomainId> {
+
+            public:
+
+                // member types
+                using communicator_type = Communicator;
+                using address_type = typename communicator_type::address_type;
+                using index_type = Index;
+                using grid_type = unstructured::detail::grid<index_type>;
+                using domain_id_type = DomainId;
+                using pattern_container_type = pattern_container<communicator_type, grid_type, domain_id_type>;
+
+                friend class pattern_container<communicator_type, grid_type, domain_id_type>;
+
+                /** @brief essentially a partition index and a sequence of remote indexes;
+                 * number of levels is provided as well, defaul is 1 (2D case):
+                 * the assumption is that each 2D element is a column, with 'levels' vertical elements.
+                 * WARN: a second index, and therefore a more complex iteration space,
```

Comment:
warning not needed?

Reply:
Warning removed, documentation is updated